src/app/modules/sns/sns.service.ts

```diff
@@ -0,0 +1,224 @@
+// 1) If the old bounce document indicates that an email bounced, set hasBounced
+// to true for that email.
+// 2) If the new delivery notification indicates that an email was delivered
+// successfully, set hasBounced to false for that email, even if the old bounce
+// document indicates that that email previously bounced.
+// 3) Update the old recipient list according to the newest bounce notification.
+const updateBounceDoc = (
+  oldBounces: IBounceSchema,
+  latestBounces: IBounceSchema,
+  snsInfo: IEmailNotification,
+): void => {
+  const isDelivery = isDeliveryNotification(snsInfo)
+  oldBounces.bounces.forEach((oldBounce) => {
+    // If we were previously notified that a given email has bounced,
+    // we want to retain that information
+    if (oldBounce.hasBounced) {
+      // Check if the latest recipient list contains that email
+      const matchedLatestBounce = latestBounces.bounces.find(
+        (newBounce) => newBounce.email === oldBounce.email,
+      )
+      // Check if the latest notification indicates that this email
+      // actually succeeded
+      const hasSubsequentlySucceeded =
+        isDelivery &&
+        get(snsInfo, 'delivery.recipients').includes(oldBounce.email)
+      if (matchedLatestBounce) {
+        // Set the latest bounce status based on the latest notification
+        matchedLatestBounce.hasBounced = !hasSubsequentlySucceeded
+      }
+    }
+  })
+  oldBounces.bounces = latestBounces.bounces
+}
+
+// Writes a log message if all recipients have bounced
+const logCriticalBounce = (bounceInfo: IBounceSchema, formId: string): void => {
+  if (
+    !bounceInfo.hasAlarmed &&
+    bounceInfo.bounces.every((emailInfo) => emailInfo.hasBounced)
+  ) {
+    logger.warn({
+      type: 'CRITICAL BOUNCE',
+      formId,
+      recipients: bounceInfo.bounces.map((emailInfo) => emailInfo.email),
+    })
+    // We don't want a flood of logs and alarms, so we use this to limit the rate of
+    // critical bounce logs for each form ID
+    bounceInfo.hasAlarmed = true
+  }
+}
+
+/**
+ * Parses an SNS notification and updates the Bounce collection.
+ * @param body The request body of the notification
+ */
+export const updateBounces = async (body: ISnsNotification): Promise<void> => {
+  const notification: IEmailNotification = JSON.parse(body.Message)
+  logger.info(notification)
+  const emailType = extractHeader(notification, EMAIL_HEADERS.emailType)
+  const formId = extractHeader(notification, EMAIL_HEADERS.formId)
+  // We only care about admin emails
+  if (emailType !== EMAIL_TYPES.adminResponse || !formId) return
+  const latestBounces = extractBounceDoc(notification, formId)
+  const oldBounces = await Bounce.findOne({ formId })
+  if (oldBounces) {
+    updateBounceDoc(oldBounces, latestBounces, notification)
+    logCriticalBounce(oldBounces, formId)
+    await oldBounces.save()
+  } else {
+    logCriticalBounce(latestBounces, formId)
+    await latestBounces.save()
+  }
```

Comment:
What do you think about extending the _domain model_ so that you can achieve APIs that look like the following?

```typescript
  const latestBounces = Bounce.fromNotification(notification, formId)
  const oldBounces = await Bounce.findOne({ formId })
  if (oldBounces) {
    oldBounces.update(latestBounces, notification)
    logCriticalBounce(oldBounces, formId)
    await oldBounces.save()
  } else {
    logCriticalBounce(latestBounces, formId)
    await latestBounces.save()
  }
```

Reply:
Hmm I think this could go either way. Extending the domain model helps make the service cleaner, but it also means that knowledge about the format of notifications is spread out across the codebase rather than contained in one service, e.g. code such as `get(snsInfo, 'mail.commonHeaders.to')` will have to be moved to the Bounce model. I'll take your suggestion though.

Reply:
On second thoughts if you're okay with it, I think it's a good idea to change `updateBounceDoc` to a schema method, but I think `extractBounceDoc` should stay in `sns.service` so we only have to worry about the format of SNS notifications in one place.

Reply:
On second thoughts I'd rather keep the Bounce schema agnostic to the specifics of SNS notifications, and keep all the parsing code in the SNS service. If we extend the domain model then both `Bounce.fromnNotification` and `bounce.update` will need to know about SNS notifications in order to parse them. What do you think?

Reply:
@liangyuanruo bumping this for your thoughts on the above! if you're ok then I'll merge it, if not I'm okay to refactor as well

Reply:
Thanks for being open, yes please refactor. If you prefer it to be defined outside the schema, you can introduce a _BounceFactory_ instead.

On your point on the update function - we should have the ability for one bounce object to update another, instead of using a notification object. Something like

```typescript
const bounce1 = BounceFactory.fromNotification(notif1)
const bounce2 = BounceFactory.fromNotification(notif2)
bounce1.merge(bounce2)
```